src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs

```diff
@@ -4632,37 +4699,195 @@ override public Task<T> GetFieldValueAsync<T>(int i, CancellationToken cancellat
                 return source.Task;
             }
 
-            // Setup cancellations
             IDisposable registration = null;
             if (cancellationToken.CanBeCanceled)
             {
-                registration = cancellationToken.Register(s => ((SqlCommand)s).CancelIgnoreFailure(), _command);
+                registration = cancellationToken.Register(SqlCommand.s_cancelIgnoreFailure, _command);
             }
 
-            // Setup async
             PrepareAsyncInvocation(useSnapshot: true);
 
-            // Setup the retryable function
-            Func<Task, Task<T>> moreFunc = null;
-            moreFunc = (t) =>
-            {
-                if (t != null)
+            return InvokeResumable(
+                new GetFieldValueAsyncResumable<T>(this, source, registration)
                 {
-                    PrepareForAsyncContinuation();
+                    columnIndex = i
                 }
+            );
+        }
 
-                if (TryReadColumn(i, setTimeout: false))
-                {
-                    return Task.FromResult<T>(GetFieldValueFromSqlBufferInternal<T>(_data[i], _metaData[i]));
-                }
-                else
-                {
-                    return ContinueRetryable(moreFunc);
-                }
-            };
+        private static Task<T> GetFieldValueAsyncWillResume<T>(Task task, object state)
+        {
+            GetFieldValueAsyncResumable<T> resumable = (GetFieldValueAsyncResumable<T>)state;
+            SqlDataReader reader = resumable.reader;
+            int columnIndex = resumable.columnIndex;
+
+            if (task != null)
+            {
+                reader.PrepareForAsyncContinuation();
+            }
+
+            if (reader.TryReadColumn(columnIndex, setTimeout: false))
+            {
+                return Task.FromResult<T>(reader.GetFieldValueFromSqlBufferInternal<T>(reader._data[columnIndex], reader._metaData[columnIndex]));
+            }
+            else
+            {
+                return reader.ContinueResumable(resumable);
+            }
+        }
+
+
+        private abstract class Resumable<T> : IDisposable
+        {
+            internal static readonly Lazy<Action<Task<T>, object>> s_completeCallback = new Lazy<Action<Task<T>, object>>(
```

Comment:
Why Lazy?

Reply:
Because it may not be needed if there's a failure. It could be non-lazy if it's preferred, I don't feel strongly about it.

Reply:
Only saves 24 bytes for a static
```
Type layout for 'Action`2'
Size: 48 bytes. Paddings: 0 bytes (%0 of empty space)
|==========================================|
| Object Header (8 bytes)                  |
|------------------------------------------|
| Method Table Ptr (8 bytes)               |
|==========================================|
|   0-7: Object _target (8 bytes)          |
|------------------------------------------|
|  8-15: Object _methodBase (8 bytes)      |
|------------------------------------------|
| 16-23: IntPtr _methodPtr (8 bytes)       |
|------------------------------------------|
| 24-31: IntPtr _methodPtrAux (8 bytes)    |
|------------------------------------------|
| 32-39: Object _invocationList (8 bytes)  |
|------------------------------------------|
| 40-47: IntPtr _invocationCount (8 bytes) |
|==========================================|
```
```
Type layout for 'Action`2'
Size: 48 bytes. Paddings: 0 bytes (%0 of empty space)
|==========================================|
| Object Header (8 bytes)                  |
|------------------------------------------|
| Method Table Ptr (8 bytes)               |
|==========================================|
|   0-7: Object _target (8 bytes)          |
|------------------------------------------|
|  8-15: Object _methodBase (8 bytes)      |
|------------------------------------------|
| 16-23: IntPtr _methodPtr (8 bytes)       |
|------------------------------------------|
| 24-31: IntPtr _methodPtrAux (8 bytes)    |
|------------------------------------------|
| 32-39: Object _invocationList (8 bytes)  |
|------------------------------------------|
| 40-47: IntPtr _invocationCount (8 bytes) |
|==========================================|
```

Reply:
It does seem like we should be optimizing for the success path - the failure path is likely to be expensive in various ways in other cases (e.g. exception handling).

Reply:
Ok, done.